// @vueuse/firebase/index.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vueDemi = require('vue-demi');
var shared = require('@vueuse/shared');

function useAuth(authInstance) {
  let auth;
  if (authInstance instanceof Function)
    auth = authInstance();
  else
    auth = authInstance;
  const user = vueDemi.ref(auth.currentUser);
  const isAuthenticated = vueDemi.computed(() => !!user.value);
  auth.onIdTokenChanged((authUser) => user.value = authUser);
  return {
    isAuthenticated,
    user
  };
}

function getData(docRef) {
  const data = docRef.data();
  if (data) {
    Object.defineProperty(data, "id", {
      value: docRef.id.toString(),
      writable: false
    });
  }
  return data;
}
function isDocumentReference(docRef) {
  var _a;
  return (((_a = docRef.path) == null ? void 0 : _a.match(/\//g)) || []).length % 2 !== 0;
}
function useFirestore(docRef, initialValue = void 0, options = {}) {
  const {
    errorHandler = (err) => console.error(err),
    autoDispose = true
  } = options;
  if (isDocumentReference(docRef)) {
    const data = vueDemi.ref(initialValue);
    const close = docRef.onSnapshot((snapshot) => {
      data.value = getData(snapshot) || null;
    }, errorHandler);
    shared.tryOnScopeDispose(() => {
      close();
    });
    return data;
  } else {
    const data = vueDemi.ref(initialValue);
    const close = docRef.onSnapshot((snapshot) => {
      data.value = snapshot.docs.map(getData).filter(shared.isDef);
    }, errorHandler);
    if (autoDispose) {
      shared.tryOnScopeDispose(() => {
        close();
      });
    }
    return data;
  }
}

function useRTDB(docRef, options = {}) {
  const {
    autoDispose = true
  } = options;
  const data = vueDemi.ref(void 0);
  function update(snapshot) {
    data.value = snapshot.val();
  }
  docRef.on("value", update);
  if (autoDispose) {
    shared.tryOnScopeDispose(() => {
      docRef.off("value", update);
    });
  }
  return data;
}

exports.useAuth = useAuth;
exports.useFirestore = useFirestore;
exports.useRTDB = useRTDB;
