// @vueuse/firebase/index.mjs
import { ref, computed } from 'vue-demi';
import { tryOnScopeDispose, isDef } from '@vueuse/shared';

function useAuth(authInstance) {
  let auth;
  if (authInstance instanceof Function)
    auth = authInstance();
  else
    auth = authInstance;
  const user = ref(auth.currentUser);
  const isAuthenticated = computed(() => !!user.value);
  auth.onIdTokenChanged((authUser) => user.value = authUser);
  return {
    isAuthenticated,
    user
  };
}

function getData(docRef) {
  const data = docRef.data();
  if (data) {
    Object.defineProperty(data, "id", {
      value: docRef.id.toString(),
      writable: false
    });
  }
  return data;
}
function isDocumentReference(docRef) {
  var _a;
  return (((_a = docRef.path) == null ? void 0 : _a.match(/\//g)) || []).length % 2 !== 0;
}
function useFirestore(docRef, initialValue = void 0, options = {}) {
  const {
    errorHandler = (err) => console.error(err),
    autoDispose = true
  } = options;
  if (isDocumentReference(docRef)) {
    const data = ref(initialValue);
    const close = docRef.onSnapshot((snapshot) => {
      data.value = getData(snapshot) || null;
    }, errorHandler);
    tryOnScopeDispose(() => {
      close();
    });
    return data;
  } else {
    const data = ref(initialValue);
    const close = docRef.onSnapshot((snapshot) => {
      data.value = snapshot.docs.map(getData).filter(isDef);
    }, errorHandler);
    if (autoDispose) {
      tryOnScopeDispose(() => {
        close();
      });
    }
    return data;
  }
}

function useRTDB(docRef, options = {}) {
  const {
    autoDispose = true
  } = options;
  const data = ref(void 0);
  function update(snapshot) {
    data.value = snapshot.val();
  }
  docRef.on("value", update);
  if (autoDispose) {
    tryOnScopeDispose(() => {
      docRef.off("value", update);
    });
  }
  return data;
}

export { useAuth, useFirestore, useRTDB };
